zebra-state/src/service/finalized_state.rs

```diff
@@ -346,14 +353,40 @@ impl FinalizedState {
                 }
             }
 
+            let sapling_root = sapling_note_commitment_tree.root();
+            let orchard_root = orchard_note_commitment_tree.root();
+
+            // Create the history tree if it's the Heartwood activation block.
+            let heartwood_height = NetworkUpgrade::Heartwood
+                .activation_height(self.network)
+                .expect("Heartwood height is known");
+            match height.cmp(&heartwood_height) {
+                std::cmp::Ordering::Less => assert!(history_tree.is_none()),
+                std::cmp::Ordering::Equal => {
+                    history_tree = Some(HistoryTree::from_block(
+                        self.network,
+                        block.clone(),
+                        &sapling_root,
+                        &orchard_root,
+                    )?);
+                }
+                std::cmp::Ordering::Greater => assert!(history_tree.is_some()),
+            }
+
+            // Add block to history tree if applicable
+            if let Some(history_tree) = &mut history_tree {
+                history_tree.push(block.clone(), &sapling_root, &orchard_root)?;
```

Comment:
Confusing log message:

Since we're adding a non-disk error here, we should also update this `expect` message:
https://github.com/ZcashFoundation/zebra/blob/4f96a4bb403ac23351446230fcdb2016954ee884/zebra-state/src/service.rs#L163

(The non-finalized state has already checked the history, so it's ok for it to `expect` no history and disk errors.)

Reply:
Good catch. I returned to the previous behavior as explained in the other comment, so this no longer applies

Reply:
Unfortunately, I don't think we can avoid returning errors, so we need to change this `expect` message.